src/oc_erchef/schema/deploy/create_and_update_users.sql

```diff
@@ -0,0 +1,105 @@
+-- Deploy enterprise_chef:create_and_update_users to pg
+-- Inserts a sentinel value into the public_key field of a user table.
+-- Ensures keys table is the only source of truth for public_key
+
+BEGIN;
+
+DROP FUNCTION If EXISTS add_user(character,character,text,text,text,integer,text,text,password_hash_type,character,timestamp without time zone,timestamp without time zone,text,boolean,text,boolean);
+
+CREATE OR REPLACE FUNCTION add_user(p_id users.id%TYPE,
+                                    p_authz_id users.authz_id%TYPE,
+                                    p_username users.username%TYPE,
+                                    p_email users.email%TYPE,
+                                    p_public_key users.public_key%TYPE,
+                                    p_pubkey_version users.pubkey_version%TYPE,
+                                    p_hashed_password users.hashed_password%TYPE,
+                                    p_salt users.salt%TYPE,
+                                    p_hash_type users.hash_type%TYPE,
+                                    p_last_updated_by users.last_updated_by%TYPE,
+                                    p_created_at users.created_at%TYPE,
+                                    p_updated_at users.updated_at%TYPE,
+                                    p_external_authentication_uid users.external_authentication_uid%TYPE,
+                                    p_recovery_authentication_enabled users.recovery_authentication_enabled%TYPE,
+                                    p_serialized_object users.serialized_object%TYPE,
+                                    p_admin users.admin%TYPE) RETURNS integer AS $$
+  DECLARE
+      inserteduser integer;
+  BEGIN
+   -- Need this insert despite of the trigger since the trigger updates the keys table
+   -- from the value inserted in the users table. But we insert sentinel in the users table.
+    IF p_public_key != 'this_is_not_a_key' THEN
+      INSERT INTO keys
+        (id, key_name, public_key, key_version, created_at,  expires_at)
+      VALUES
+         (p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp);
+    END IF;
+    WITH createduser AS
+      (INSERT INTO users
+        (id, authz_id, username, email, public_key, hashed_password, salt, hash_type,
+         last_updated_by, created_at, updated_at, external_authentication_uid,
+         recovery_authentication_enabled, serialized_object, admin,
+         pubkey_version)
+       VALUES (p_id, p_authz_id, p_username, p_email, 'this_is_not_a_key',
```

Comment:
If the intention is to have the `keys` table be the source of truth for keys, why keep key-related information in the `users` table at all? Is this step 1 of a broader refactoring?

Reply:
yep - sort of a phase1 since on env like hosted we need to support older versions of chef-server.